In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
from sklearn.metrics import precision_score, recall_score, accuracy_score, f1_score
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import confusion_matrix

In [2]:
# https://www.kaggle.com/mlg-ulb/creditcardfraud

df = pd.read_csv("creditcard.csv")
df.dropna(axis = 0, inplace=True)
df.drop(['Time'], axis = 1, inplace=True)
df.head()

V1        V2        V3        V4        V5        V6        V7  \
0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9       V10  ...       V21       V22       V23       V24  \
0  0.098698  0.363787  0.090794  ... -0.018307  0.277838 -0.110474  0.066928   
1  0.085102 -0.255425 -0.166974  ... -0.225775 -0.638672  0.101288 -0.339846   
2  0.247676 -1.514654  0.207643  ...  0.247998  0.771679  0.909412 -0.689281   
3  0.377436 -1.387024 -0.054952  ... -0.108300  0.005274 -0.190321 -1.175575   
4 -0.270533  0.817739  0.753074  ... -0.009431  0.798278 -0.137458  0.141267   

        V25       V26       V27       V28  Amount  Class  
0  0.128539 -0.189115  0.133558 -0.021053  149.62    0.0  
1  0.167170  0.125895 -0.008983  0.014724    2.69    0.0  
2 -0.327642 -0.139097 -0.055353 -0.059752  378.66    0.0  
3  0.647376 -0.221929  0.062723  0.061458  123.50    0.0  
4 -0.206010  0.502292  0.219422  0.215153   69.99    0.0  

[5 rows x 30 columns]

In [3]:
df.describe()

V1            V2            V3            V4            V5  \
count  25837.000000  25837.000000  25837.000000  25837.000000  25837.000000   
mean      -0.225712      0.166964      0.723131      0.229368     -0.194354   
std        1.889067      1.552825      1.684909      1.434242      1.436206   
min      -30.552380    -40.978852    -31.103685     -5.172595    -42.147898   
25%       -0.951371     -0.405182      0.279675     -0.683788     -0.779682   
50%       -0.268134      0.176391      0.862364      0.204882     -0.224921   
75%        1.166445      0.818213      1.489349      1.109950      0.320600   
max        1.960497     16.713389      4.101716     13.143668     34.099309   

                 V6            V7            V8            V9           V10  \
count  25837.000000  25837.000000  25837.000000  25837.000000  25837.000000   
mean       0.079800     -0.128669      0.026060      0.441922     -0.144162   
std        1.323018      1.319114      1.341622      1.269785      1.183321   
min      -23.496714    -26.548144    -41.484823     -7.175097    -14.166795   
25%       -0.662365     -0.594982     -0.164442     -0.384700     -0.619551   
50%       -0.179636     -0.067756      0.029393      0.358908     -0.197733   
75%        0.466671      0.446339      0.287518      1.200785      0.296187   
max       22.529298     36.677268     20.007208     10.392889     12.701539   

       ...           V21           V22           V23           V24  \
count  ...  25837.000000  25837.000000  25837.000000  25837.000000   
mean   ...     -0.037598     -0.132262     -0.039778      0.011174   
std    ...      0.817316      0.641426      0.538789      0.590048   
min    ...    -20.262054     -8.593642    -26.751119     -2.836627   
25%    ...     -0.249309     -0.552080     -0.173949     -0.325263   
50%    ...     -0.097962     -0.106938     -0.049040      0.063000   
75%    ...      0.072314      0.276513      0.075082      0.399042   
max    ...     22.614889      5.805795     13.876221      4.014444   

                V25           V26           V27           V28        Amount  \
count  25837.000000  25837.000000  25837.000000  25837.000000  25837.000000   
mean       0.129413      0.023418      0.013306      0.004153     75.808937   
std        0.436213      0.516670      0.397674      0.280120    219.661154   
min       -7.495741     -1.338556     -8.567638     -9.617915      0.000000   
25%       -0.133337     -0.338941     -0.065269     -0.008898      6.060000   
50%        0.167170     -0.052787      0.007047      0.019985     18.960000   
75%        0.410283      0.306226      0.091845      0.076753     67.610000   
max        5.525093      3.517346     11.135740      4.860769   7879.420000   

              Class  
count  25837.000000  
mean       0.003406  
std        0.058262  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000  

[8 rows x 30 columns]

In [4]:
df.Class.value_counts()

0.0    25749
1.0       88
Name: Class, dtype: int64

In [5]:
X_train, X_test, y_train, y_test = train_test_split(df.drop("Class", axis=1), df.Class, train_size=.7, random_state=42)


In [6]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

Vamos a calcular la exactitud de Regresión Logística:

In [7]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()
model.fit(X_train_scaled, y_train)
y_pred = model.predict(X_test_scaled)
print(accuracy_score(y_test, y_pred))

0.9979360165118679


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Si examinamos con más detalle otras métricas vemos que no es tan buena:

In [8]:
conf = confusion_matrix(y_test, y_pred)
conf

array([[7718,    6],
       [  10,   18]])

In [9]:
print("Accuracy: ", accuracy_score(y_test, y_pred))
print("Recall: ", recall_score(y_test, y_pred))
print("precision_score: ", precision_score(y_test, y_pred))
print("Especificidad: ", recall_score(y_test, y_pred, pos_label=0)) # Negativos predichos correctamente
print("F1-score: ", f1_score(y_test, y_pred))

Accuracy:  0.9979360165118679
Recall:  0.6428571428571429
precision_score:  0.75
Especificidad:  0.9992232004142931
F1-score:  0.6923076923076924


De las transacciones fraudulentas, aunque pocás, ha acertado poco menos de la mitad

Veamos si la detección de anomalías funciona un poco mejor:

**PREGUNTA**
1. Utiliza el Determinante de Covarianza Mínima para detectar los outliers y hacer predicciones con la misma. Calcula todas las métricas anteriores y muestra la matriz de confusión

In [20]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import confusion_matrix, accuracy_score, recall_score, precision_score, f1_score
from sklearn.covariance import EllipticEnvelope

df = pd.read_csv("creditcard.csv")
df.dropna(axis=0, inplace=True)
df.drop(['Time'], axis=1, inplace=True)

# Separo las características (X) y la variable objetivo (y)
X_train, X_test, y_train, y_test = train_test_split(df.drop("Class", axis=1), df.Class, train_size=.7, random_state=42)

# Estandarizo las características
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Entreno el modelo de Detección de Anomalías (Elliptic Envelope)
anomaly_detector = EllipticEnvelope(contamination=0.02)
anomaly_detector.fit(X_train_scaled)



/usr/local/lib/python3.10/dist-packages/sklearn/covariance/_robust_covariance.py:184: RuntimeWarning: Determinant has increased; this should not happen: log(det) > log(previous_det) (-242.456352107325870 > -244.348661575689164). You may want to try with a higher value of support_fraction (current value: 0.501).
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/covariance/_robust_covariance.py:184: RuntimeWarning: Determinant has increased; this should not happen: log(det) > log(previous_det) (-242.968368758560530 > -246.483433123917735). You may want to try with a higher value of support_fraction (current value: 0.501).
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/covariance/_robust_covariance.py:184: RuntimeWarning: Determinant has increased; this should not happen: log(det) > log(previous_det) (-237.158303907396999 > -243.614575763711002). You may want to try with a higher value of support_fraction (current value: 0.501).
  warnings.warn(
/usr/local

EllipticEnvelope(contamination=0.02)

In [21]:
# Realizo predicciones en el conjunto de prueba
y_pred_anomaly = anomaly_detector.predict(X_test_scaled)

# Convierto las predicciones a etiquetas binarias (1 para normal, -1 para anomalía)
y_pred_anomaly_binary = [1 if pred == 1 else 0 for pred in y_pred_anomaly]


# Matriz de confusión
conf_anomaly = confusion_matrix(y_test, y_pred_anomaly_binary)
print("Matriz de Confusión:\n", conf_anomaly)


print("Accuracy: ", accuracy_score(y_test, y_pred_anomaly_binary))
print("Recall: ", recall_score(y_test, y_pred_anomaly_binary))
print("Precision: ", precision_score(y_test, y_pred_anomaly_binary))
print("Specificity: ", recall_score(y_test, y_pred_anomaly_binary, pos_label=0))
print("F1-score: ", f1_score(y_test, y_pred_anomaly_binary))



Matriz de Confusión:
 [[ 1652 83655]
 [   28   108]]
Accuracy:  0.020598527673419707
Recall:  0.7941176470588235
Precision:  0.001289352100569464
Specificity:  0.01936535102629327
F1-score:  0.002574524130204174


las metricas que he obtenido como resultado de la detección de anomalías utilizando el Determinante de Covarianza Mínima (MCD) indican un rendimiento bastante bajo en este conjunto de datos.

Accuracy (Exactitud): 0.0206

La exactitud es muy baja, lo que sugiere que el modelo tiene dificultades para clasificar correctamente las transacciones normales y fraudulentas.

Recall (Recuperación): 0.7941

La tasa de recuperación es relativamente alta, lo que indica que el modelo ha identificado correctamente casi el 80% de las transacciones fraudulentas.

Precision (Precisión): 0.0013

La precisión es extremadamente baja, indicando que la gran mayoría de las transacciones que el modelo clasifica como fraudulentas son falsos positivos.

Specificity (Especificidad): 0.0194

La especificidad es baja, indicando que el modelo tiene dificultades para identificar correctamente las transacciones normales.

F1-score: 0.0026

El F1-score, que combina precisión y recall, también es muy bajo.

Matriz de Confusión:
[[  499, 50528],
 [   57,  55]]

El modelo ha identificado 108 transacciones fraudulentas correctamente (verdaderos positivos) pero ha clasificado erróneamente 83655 transacciones normales como anomalías (falsos positivos).

Estos resultados sugieren que, aunque el modelo tiene un alto recall (buena capacidad para identificar transacciones fraudulentas), tiene dificultades para mantener la precisión y especificidad, clasificando muchas transacciones normales como anomalías (falsos positivos)



**PREGUNTA**
2. Utiliza el Isolation Forest para detectar los outliers y hacer predicciones con el mismo. Calcula las métricas y muestra la matriz de confusión

In [22]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import IsolationForest
from sklearn.metrics import confusion_matrix, accuracy_score, recall_score, precision_score, f1_score

df = pd.read_csv("creditcard.csv")
df.dropna(axis=0, inplace=True)
df.drop(['Time'], axis=1, inplace=True)

X_train, X_test, y_train, y_test = train_test_split(df.drop("Class", axis=1), df.Class, train_size=.7, random_state=42)

# Estandarizo las características
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Entreno el modelo Isolation Forest
isolation_forest = IsolationForest(contamination=0.02, random_state=42)
isolation_forest.fit(X_train_scaled)

# Realizo predicciones en el conjunto de prueba
y_pred_isolation = isolation_forest.predict(X_test_scaled)

# Convierto las predicciones a etiquetas binarias (1 para normal, -1 para anomalía)
y_pred_isolation_binary = [1 if pred == 1 else 0 for pred in y_pred_isolation]

# Muestro la matriz de confusión
conf_isolation = confusion_matrix(y_test, y_pred_isolation_binary)
print("Confusion Matrix:\n", conf_isolation)

print("Accuracy: ", accuracy_score(y_test, y_pred_isolation_binary))
print("Recall: ", recall_score(y_test, y_pred_isolation_binary))
print("Precision: ", precision_score(y_test, y_pred_isolation_binary))
print("Specificity: ", recall_score(y_test, y_pred_isolation_binary, pos_label=0))
print("F1-score: ", f1_score(y_test, y_pred_isolation_binary))




Confusion Matrix:
 [[ 1549 83758]
 [  105    31]]
Accuracy:  0.018491860070456327
Recall:  0.22794117647058823
Precision:  0.00036997696595018436
Specificity:  0.018157947179012274
F1-score:  0.0007387548406315162


Matriz de Confusión:
[[ 1549, 83758],
 [  105,    31]]

El modelo ha identificado 31 transacciones fraudulentas correctamente (verdaderos positivos), pero ha clasificado erróneamente 83758 transacciones normales como anomalías (falsos positivos).

Accuracy (Exactitud): 0.0185

La exactitud sigue siendo muy baja, indicando que el modelo tiene dificultades para clasificar correctamente las transacciones normales y fraudulentas.

Recall (Recuperación): 0.2279

La tasa de recuperación ha disminuido significativamente, lo que indica que el modelo ahora identifica menos transacciones fraudulentas.

Precision (Precisión): 0.00037

La precisión sigue siendo extremadamente baja, indicando que la gran mayoría de las transacciones que el modelo clasifica como fraudulentas son falsos positivos.

Specificity (Especificidad): 0.0182

La especificidad también ha disminuido, indicando que el modelo sigue teniendo dificultades para identificar correctamente las transacciones normales.

F1-score: 0.00074

El F1-score sigue siendo muy bajo.


**PREGUNTA**
3. ¿Qué porcentaje de predicciones como positivas eran correctas?
4. ¿Qué porcentaje de casos que eran positivos se han predicho correctamente?
5. ¿Qué porcentaje de casos que eran negativos se han predicho correctamente?

3.¿Qué porcentaje de predicciones como positivas eran correctas?

El porcentaje de predicciones positivas que son correctas (la precisión) es aproximadamente del 0.04%. Es una tasa extremadamente baja, lo que indica que el modelo tiene dificultades para predecir correctamente las transacciones fraudulentas sin clasificar muchas transacciones normales como fraudulentas (falsos positivos).

Precision = 31(VP) / 31(VP) + 83758(FP)**texto en negrita** = 0.00036976965

4.¿Qué porcentaje de casos que eran positivos se han predicho correctamente?

El porcentaje de casos que eran positivos y que fueron predichos correctamente (recall) es aproximadamente del 22.79%. Esto indica que el modelo identificó correctamente alrededor del 22.79% de las transacciones fraudulentas presentes en el conjunto de prueba.

Recall = 31(VP) / 31(VP) + 105(FN) = 0.2279411764705

5. ¿Qué porcentaje de casos que eran negativos se han predicho correctamente?

El porcentaje de casos que eran negativos y que fueron predichos correctamente (especificidad) es aproximadamente del 98.18%. Esto indica que el modelo identificó correctamente alrededor del 98.18% de las transacciones normales presentes en el conjunto de prueba.

Especificidad = 83758(VN) / 83758(VN) + 1549(FP)= 0.98184205283


**PREGUNTA**

6. Utiliza Logistic Regression sensible al coste y comprueba las métricas. ¿Ha mejorado el Recall? Imprime la matriz de confusión
7. Haz lo mismo haciendo undersampling
8. Haz lo mismo con oversampling
9. Haz lo mismo con SMOTE

6.Utiliza Logistic Regression sensible al coste y comprueba las métricas. ¿Ha mejorado el Recall? Imprime la matriz de confusión

In [23]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, recall_score

# Defino el modelo de Regresión Logística
model_lr_cost_sensitive = LogisticRegression(class_weight='balanced', random_state=42)

# Entreno el modelo
model_lr_cost_sensitive.fit(X_train_scaled, y_train)

# Realizo las predicciones en el conjunto de prueba
y_pred_lr_cost_sensitive = model_lr_cost_sensitive.predict(X_test_scaled)

# Calculo de las métricas
accuracy_lr_cost_sensitive = accuracy_score(y_test, y_pred_lr_cost_sensitive)
recall_lr_cost_sensitive = recall_score(y_test, y_pred_lr_cost_sensitive)
confusion_lr_cost_sensitive = confusion_matrix(y_test, y_pred_lr_cost_sensitive)

print(f'Accuracy del modelo de Regresión Logística sensible al coste: {accuracy_lr_cost_sensitive}')
print(f'Recall del modelo de Regresión Logística sensible al coste: {recall_lr_cost_sensitive}')
print(f'\nMatriz de Confusión:')
print(confusion_lr_cost_sensitive)


Accuracy del modelo de Regresión Logística sensible al coste: 0.9742167292815093
Recall del modelo de Regresión Logística sensible al coste: 0.9264705882352942

Matriz de Confusión:
[[83114  2193]
 [   10   126]]


El recall anterior era de aproximadamente 22.79%, y después de aplicar la sensibilidad al costo, el recall ha mejorado considerablemente a aproximadamente 92.65%.

Se puede decir que el modelo sensible al costo y está identificando una proporción mucho mayor de las transacciones fraudulentas presentes en el conjunto de prueba, lo cual es un resultado positivo.

La matriz de confusión también muestra un aumento en los verdaderos positivos (transacciones fraudulentas correctamente identificadas) y una disminución en los falsos negativos.

7.Haz lo mismo haciendo undersampling

In [25]:
from imblearn.under_sampling import RandomUnderSampler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, recall_score
from sklearn.preprocessing import StandardScaler

# Realizo undersampling en el conjunto de entrenamiento
undersampling_train = RandomUnderSampler(random_state=42)
X_train_resampled, y_train_resampled = undersampling_train.fit_resample(X_train_scaled, y_train)

# Defino el modelo de Regresión Logística
model_lr_undersampling = LogisticRegression(class_weight='balanced', random_state=42)

# Entreno el modelo con el conjunto de entrenamiento undersampled
model_lr_undersampling.fit(X_train_resampled, y_train_resampled)

# Realizo las predicciones en el conjunto de prueba
y_pred_lr_undersampling = model_lr_undersampling.predict(X_test_scaled)

# Calculo todas las métricas
accuracy_lr_undersampling = accuracy_score(y_test, y_pred_lr_undersampling)
recall_lr_undersampling = recall_score(y_test, y_pred_lr_undersampling)
confusion_lr_undersampling = confusion_matrix(y_test, y_pred_lr_undersampling)

print(f'Accuracy del modelo de Regresión Logística con undersampling: {accuracy_lr_undersampling}')
print(f'Recall del modelo de Regresión Logística con undersampling: {recall_lr_undersampling}')
print(f'\nMatriz de Confusión:')
print(confusion_lr_undersampling)


Accuracy del modelo de Regresión Logística con undersampling: 0.9608862048383132
Recall del modelo de Regresión Logística con undersampling: 0.9338235294117647

Matriz de Confusión:
[[81974  3333]
 [    9   127]]


8.Haz lo mismo con oversampling

In [30]:
from imblearn.over_sampling import RandomOverSampler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, recall_score
from sklearn.preprocessing import StandardScaler

# Realizo oversampling en el conjunto de entrenamiento
oversampling_train = RandomOverSampler(random_state=42)
X_train_oversampled, y_train_oversampled = oversampling_train.fit_resample(X_train_scaled, y_train)

# Defino el modelo de Regresión Logística
model_lr_oversampling = LogisticRegression(class_weight='balanced', random_state=42)

# Entreno el modelo con el conjunto de entrenamiento oversampled
model_lr_oversampling.fit(X_train_oversampled, y_train_oversampled)

# Realizo predicciones en el conjunto de prueba
y_pred_lr_oversampling = model_lr_oversampling.predict(X_test_scaled)

# Calculo de las métricas
accuracy_lr_oversampling = accuracy_score(y_test, y_pred_lr_oversampling)
recall_lr_oversampling = recall_score(y_test, y_pred_lr_oversampling)
confusion_lr_oversampling = confusion_matrix(y_test, y_pred_lr_oversampling)

print(f'Accuracy del modelo de Regresión Logística con oversampling: {accuracy_lr_oversampling}')
print(f'Recall del modelo de Regresión Logística con oversampling: {recall_lr_oversampling}')
print(f'\nMatriz de Confusión:')
print(confusion_lr_oversampling)


Accuracy del modelo de Regresión Logística con oversampling: 0.9740645810657398
Recall del modelo de Regresión Logística con oversampling: 0.9264705882352942

Matriz de Confusión:
[[83101  2206]
 [   10   126]]


9.Haz lo mismo con SMOTE



In [32]:
from imblearn.over_sampling import SMOTE
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, recall_score
from sklearn.preprocessing import StandardScaler

# Realizo oversampling en el conjunto de entrenamiento
smote = SMOTE(random_state=42)
X_train_oversampled, y_train_oversampled = smote.fit_resample(X_train_scaled, y_train)

# Defino el modelo de Regresión Logística
model_lr_oversampling = LogisticRegression(class_weight='balanced', random_state=42)

# Entreno el modelo con el conjunto de entrenamiento oversampled
model_lr_oversampling.fit(X_train_oversampled, y_train_oversampled)

# Realizo las predicciones en el conjunto de prueba
y_pred_lr_oversampling = model_lr_oversampling.predict(X_test_scaled)

# Calculo de las métricas
accuracy_lr_oversampling = accuracy_score(y_test, y_pred_lr_oversampling)
recall_lr_oversampling = recall_score(y_test, y_pred_lr_oversampling)
confusion_lr_oversampling = confusion_matrix(y_test, y_pred_lr_oversampling)

print(f'Accuracy del modelo de Regresión Logística con SMOTE: {accuracy_lr_oversampling}')
print(f'Recall del modelo de Regresión Logística con SMOTE: {recall_lr_oversampling}')
print(f'\nMatriz de Confusión:')
print(confusion_lr_oversampling)


Accuracy del modelo de Regresión Logística con SMOTE: 0.9724728766546118
Recall del modelo de Regresión Logística con SMOTE: 0.9264705882352942

Matriz de Confusión:
[[82965  2342]
 [   10   126]]
